File: ts/src/samchon/protocol/external/ExternalServerClientArray.ts
/// <reference path="../../API.ts" />

/// <reference path="ExternalSystemArray.ts" />

namespace samchon.protocol.external
{
	export interface IExternalServerClientArray
		extends IExternalServerArray, IExternalClientArray
	{
	}

	export abstract class ExternalServerClientArray
		extends ExternalClientArray
		implements IExternalServerClientArray
	{
		/* ---------------------------------------------------------
			CONSTRUCTORS
		--------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();
		}

		protected createChild(xml: library.XML): ExternalSystem
		{
			return this.createExternalServer(xml) as ExternalSystem;
		}
		protected abstract createExternalServer(xml: library.XML): IExternalServer;

		/* ---------------------------------------------------------
			METHOD OF CLIENT
		--------------------------------------------------------- */
		public connect(): void
		{
			for (let i: number = 0; i < this.size(); i++)
				if (this.at(i)["connect"] != undefined)
					(this.at(i) as IExternalServer).connect();
		}
	}
}

File: cpp/samchon/protocol/external/ExternalServerClientArray.cpp
#include <samchon/protocol/external/ExternalServerClientArray.hpp>

using namespace std;
using namespace samchon;
using namespace samchon::library;
using namespace samchon::protocol;
using namespace samchon::protocol::external;

ExternalServerClientArray::ExternalServerClientArray()
	: ExternalServerArray(),
	ExternalClientArray()
{
}
ExternalServerClientArray::~ExternalServerClientArray()
{
}

auto ExternalServerClientArray::createChild(shared_ptr<XML> xml) -> ExternalSystem*
{
	return createExternalServer(xml);
}